pkg/util/cert_pool.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+)
+
+// CertPool is a set of certificates.
+type certPool struct {
+	certificatesHashes map[[32]byte]struct{}
+	certificates       []*x509.Certificate
+}
+
+// NewCertPool returns a new, empty CertPool.
+func newCertPool() *certPool {
+	return &certPool{
+		certificatesHashes: make(map[[32]byte]struct{}),
+		certificates:       make([]*x509.Certificate, 0),
+	}
+}
+
+// Check if the given certificate was added to certificates bundle already
+func (cp *certPool) isCertificateDuplicate(certData []byte) bool {
+	// calculate hash sum of the given certificate
+	hash := sha256.Sum256(certData)
+
+	// check a hash existence
+	if _, ok := cp.certificatesHashes[hash]; ok {
+		return ok
+	}
+
+	// put certificate hash into a set of hashes
+	cp.certificatesHashes[hash] = struct{}{}
+
+	return false
+}
+
+// Append certificate to a pool
+func (cp *certPool) appendCertFromPEM(PEMdata []byte) error {
+	if PEMdata == nil {
+		return fmt.Errorf("certificate data can't be nil")
+	}
+
+	for {
+		var block *pem.Block
+		block, PEMdata = pem.Decode(PEMdata)
+
+		if block == nil {
+			// //If no PEM data is found,
+			// // p is nil and the whole of the input is returned in rest.
+			// if len(PEMdata) != 0 {
+			// 	return fmt.Errorf("error of decoding PEM block")
+			// }
```

Comment:
Is the out-commented code intentional? I am not very familiar with this PEM decoding, but if this code is not worth having, I think we should delete it.

Reply:
> Is the out-commented code intentional? I am not very familiar with this PEM decoding, but if this code is not worth having, I think we should delete it.

Yes of course, I will fix it.
Thank you so much for your help, hints and patience, it is very appreciated!

Reply:
> Is the out-commented code intentional? I am not very familiar with this PEM decoding, but if this code is not worth having, I think we should delete it.

Hello,
Comments were trimmed as you requested.